proposals/intraprocess_communication.md

```diff
@@ -0,0 +1,580 @@
+required, where `N` is the number of `Subscription`s. The last one will receive ownership of the published
+message, thus saving a copy.
+
+If none of the `Subscription`s want ownership of the message, `0` copies are required.
+It is possible to convert the message into a `std::shared_ptr<MessageT> msg` and to add it to every buffer.
+
+If there is 1 `Subscription` that does not want ownership while the others want it, the situation is
+equivalent to the case of everyone requesting ownership:`N-1` copies of the message are required. As before
+the last `Subscription` will receive ownership.
+
+If there is more than 1 `Subscription` that do not want ownership while the others want it, a total of `M`
+copies of the message are required, where `M` is the number of `Subscription`s that want ownership. `1` copy
+will be shared among all the `Subscription`s that do not want ownership, while `M-1` copies are for the others.
+
+
+As in the current implementation, if both inter and intra-process communication are needed, the
+`std::unique_ptr<MessageT> msg` will be converted into a `std::shared_ptr<MessageT> msg` and passed
+respectively to the `do_intra_process_publish` and `do_inter_process_publish` functions.
+
+A copy of the message will be given to all the `Subscription`s requesting ownership, while the others can
+copy the published shared pointer.
+
+The following tables show a recap of when the proposed implementation has to create a new copy of a message.
+The notation `@` indicates a memory address where the message is stored, different memory addresses correspond
+to different copies of the message.
+
+
+| publish\<T\>                     | BufferT                |    Results         |
+| -------------                    |  -----                    | -------------      |
+| unique_ptr\<MessageT\> @1        |   <ul><li>unique_ptr\<MessageT\></li></ul>   |    <ul><li>@1</li></ul>       |
+| unique_ptr\<MessageT\> @1        | <ul><li>unique_ptr\<MessageT\></li><li>unique_ptr\<MessageT\></li></ul>  |    <ul><li>@1</li><li>@2</li></ul>      |
+| unique_ptr\<MessageT\> @1        |   <ul><li>shared_ptr\<MessageT\></li></ul>   |    <ul><li>@1</li></ul>       |
+| unique_ptr\<MessageT\> @1        | <ul><li>shared_ptr\<MessageT\></li><li>shared_ptr\<MessageT\></li></ul>  |    <ul><li>@1</li><li>@1</li></ul>      |
+| unique_ptr\<MessageT\> @1        | <ul><li>unique_ptr\<MessageT\></li><li>shared_ptr\<MessageT\></li></ul>  |    <ul><li>@1</li><li>@2</li></ul>      |
+| unique_ptr\<MessageT\> @1        | <ul><li>unique_ptr\<MessageT\></li><li>shared_ptr\<MessageT\></li><li>shared_ptr\<MessageT\></li></ul>  |    <ul><li>@1</li><li>@2</li><li>@2</li></ul>      |
+| unique_ptr\<MessageT\> @1        | <ul><li>unique_ptr\<MessageT\></li><li>unique_ptr\<MessageT\></li><li>shared_ptr\<MessageT\></li><li>shared_ptr\<MessageT\></li></ul>  |    <ul><li>@1</li><li>@2</li><li>@3</li><li>@3</li></ul>      |
```

Comment:
The proposed approach with subscriber specific buffers does come with drawback though which isn't described in the current document. In the existing implementation the potential duplication of the message happens on demand when a subscriber actually takes the message. In the proposal the duplication directly at publishing time.

Considering a scenario (3) with N subscribers all taking a unique pointer. If the subscribers don't actually take the message (e.g. it busy and the message is being overwritten due to QoS settings) the existing implementation doesn't require a copy where as the proposal does result in the copy taking place anyway. So in a high load scenario the message being skipped since the subscriber is e.g. too slow still adds to the system load.

In another scenario (4) the separate buffers with the already duplicated data can't optimize the number of copies in all cases. Consider two subscribers - one taking a shared pointer and one taking a unique pointer. In the current proposal the message is always copies once. In the existing code with a single buffer serving both subscribers the logic can avoid this if e.g. the first subscriber takes the shared pointer and already gives away its ownership of the shared pointer before the second subscriber even tries to take the message. In that example there is no requirement to actually copy the data.

Reply:
One way to delay creating additional copies would be to always store shared pointers in the subscribers' buffers. Then, any additional copies (say, to covert to unique pointer) are only done when the subscriber removes the message from the buffer. This directly addresses (3).

Regarding (4), I don't see how you can avoid the copy if two subscribers want the same message but one wants a shared pointer and the other a unique pointer. Could you perhaps elaborate?

Reply:
> Could you perhaps elaborate?

As mentioned in the scenario: the first subscriber can get a const shared pointer of the existing message. After reading some data from it the subscriber releases the shared pointer. So the only reference of the message is the one in the buffer. If now the second subscriber wants to get a unique pointer the ownership can be passed from the buffer to the second subscriber. There was no need to copy the data because the usage of the message between the subscribers didn't overlap.

Reply:
@dirk-thomas sorry, but I still don't understand exactly.

The first subscriber will call the method [get](https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/mapped_ring_buffer.hpp#L106), while the second will call [pop](https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/mapped_ring_buffer.hpp#L179).

Regardless whether the ring buffer contains a unique_ptr or a shared_ptr, for the first subscriber no copies are needed and the internal message is promoted to shared_ptr.
When the second subscriber asks for the message now there is a shared_ptr inside the buffer and I don't see the mechanism where the copy is avoided

Reply:
From what I can understand, @dirk-thomas is trying to describe the following scenario (note: the temporal aspect of the scenario is important):

time_0: subscriber_a wants a const shared pointer to the message (hence, no copy)
time_1: subscriber_a is done processing the message and relinquishes ownership of the message.
time_2: subscriber_b wants a unique pointer to the message (shared pointer is converted to unique pointer, thus avoiding the copy).

Requirement: time_2 comes after time_1

Is this correct, @dirk-thomas?

Reply:
I think it is possible to do this optimization (essentially, move instead of copy for a single remaining sub) though it may not be trivial. The shared pointer counter alone can not be used as a proxy for the number of subs since the message may not have been copied to all the subscribers' buffers when this check is done, so this counter may underestimate the number of remaining subs.

Would you be okay with this being addressed as a future optimization considering it _may_ not be a very common case (large message + multiple subs with atleast one sub, say A, requiring unique ptr + A tries to get the message only after other subs are done processing the message)?

Reply:
I am not suggesting that this must be addressed. But I think this case should be described in the document and that the separate buffers make that kind of optimization more difficult.

Reply:
Fair enough.